### PR TITLE
Add buildactions support and more templates

### PR DESCRIPTION
**Pull request by hampoelz**
**Changes:**
- Use Promises with `async`/`await`
- Use vscode's `QuickPick api` to create a new file / selection template
- Some reformatting

**Added:**
- UWP Page template
- UWP UserControl template
- UWP Resource file (.resw) template
- Command to change BuildAction entry for files in project
- Set BuildAction entry in project when creating new files or adding existing files
- Change BuildAction entry in project file when renaming files / folders
- Remove BuildAction entry from project file when deleting files / folders
- `csprojWriter.ts` file to get path of project file and `set`/`get`/`remove` BuildActions from project

**Why we need this:**
If you want to use a simple, lightweight and platform-independently IDE instead of VisualStudio to develop your C# UWP-project  (like me), it is annoying to add all the project files to the .csproj or .shproj file by hand.

### CODE REVIEW

File: src/csprojWriter.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        let obj = {
+            [itemType]: {
+                $: {
+                    'Include': itemPath
+                }
+            }
+        };
+
+        if (itemType === BuildActions.Compile && itemPath.endsWith('.xaml.cs')) {
+            console.log("TRUE")
```
Comment (PoLaKoSz) [inline src/csprojWriter.ts:53]: I think `console.log`s should be removed from commits.

File: src/csprojWriter.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        itemPath = itemPath.replace(path.dirname(projPath) + path.sep, path.extname(projPath) == '.projitems' ? "$(MSBuildThisFileDirectory)" : "");
+
+        const
+            xml = await fs.readFile(projPath, 'utf8'),
+            xmlParser = util.promisify(new xml2js.Parser().parseString),
+            xmlBuilder = new xml2js.Builder();
+
+        let parsedXml = await xmlParser(xml);
+        if (parsedXml === undefined || parsedXml.Project === undefined) return;
```
Comment (PoLaKoSz) [inline src/csprojWriter.ts:42]: These lines i think repeted in the `add` and `remove` methods. Can't be moved to a method?

File: src/csprojWriter.ts (hunk 1 above)
Comment (hampoelz) [reply src/csprojWriter.ts]: Oops! Yeah, you're right

File: src/csprojWriter.ts [hunk 3]
```diff
@@ -108,15 +118,21 @@ export class CsProjWriter {
...
+                        actions.splice(action, 1);
+                    }
                 }
+
+                if (actions.length === 0) actionsArray.splice(actionsArray.indexOf(actions), 1);
             }
 
-            if (actions.length === 0) items.splice(items.indexOf(item));
+            if (actionsArray.length === 0) items.splice(items.indexOf(item), 1);
         }
```
Comment (hampoelz) [inline src/csprojWriter.ts]: Use same algorithm as with `getItemGroupByPath` to fix bugs when removing items from `ItemGroups` that contain several different `BuildActions`

**Comment by pm4rcin** [comment]
@hampoelz consider moving [here](https://github.com/Tdallau/csharpextensions) since maintainer doesn't respond. :)

**Comment by pm4rcin** [comment]
@hampoelz @PoLaKoSz maybe move [here](https://github.com/Tdallau/csharpextensions) since this fork looks like abandoned?

**Comment by hampoelz** [comment]
@pm4rcin thanks, I'll check it out

**Comment by KreativJos** [comment]
Personally I don't like the 'select template' via the top menu.
A sub menu seems like a better option, so I've tied that option in branch [feature/subMenu](https://github.com/KreativJos/csharpextensions/tree/feature/subMenu)

image

**Comment by pm4rcin** [comment]
@hampoelz since this fork went back to life would you like to rebase your code to make it usable for others? :)

**Comment by hampoelz** [comment]
@KreativJos Yeah, I agree with you, a submenu would also be easier to use.

@pm4rcin I plan to rework this Pull request later this year. I'm currently working on another project and don't have that much time.

**Comment by KreativJos** [comment]
Hi @hampoelz ! I have reworked part of your PR in PR #41 .
Would you be so kind as to take a look at that?
I'd especially would like to know why you chose the design of watching files in the extension activation function.

**Comment by hampoelz** [comment]
Nice to hear @KreativJos, I'll be checking out your PR in the next few days.

I implemented the file watcher to track rename, move, and delete actions, so that the buildaction for a file is automatically updated or removed.

**Pull request closed without merging**